internal/usde/usde.go

```diff
@@ -119,17 +123,23 @@ func resourceSpecNeedsUpdating(ctx context.Context, newDef *kubernetes.GenericOb
 			"existingIsMap", existingIsMap,
 		).Debug("checking data loss field differences")
 
-		// Note: if the data loss field is a map but it is nil in either the new or existing spec,
-		// the related variable ...IsMap will be false even though the underlying field is a
-		// map (since it is nil we do not know if it is truly a map).
-		// Therefore, the areDefFieldsMap will also be false.
-		areDefFieldsMap := newIsMap && existingIsMap
-
-		// If the current field is not explicitely a map (or if it is assumed from the config because IncludeSubfields would be true)
-		// and the config says to include comparing the subfields, then compare the fields/maps and, if the fields/maps are different,
-		// a data loss prevention strategy is needed
-		if (dataLossField.IncludeSubfields || !areDefFieldsMap) && !reflect.DeepEqual(newDefField, existingDefField) {
-			return true, upgradeStrategy, nil
+		if dataLossField.IncludeSubfields {
+			// is the definition (fields + children) at all different?
+			if !reflect.DeepEqual(newDefField, existingDefField) {
+				return true, upgradeStrategy, nil
+			}
+		} else {
+			isMap := newIsMap || existingIsMap
+			// if it's a map, since we don't care about subfields, we just need to know if it's present in one and not the other
+			if isMap {
+				if !newIsMap || !existingIsMap { // this means that one of them is nil
+					return true, upgradeStrategy, nil
+				}
+			} else {
+				if !reflect.DeepEqual(newDefField, existingDefField) {
+					return true, upgradeStrategy, nil
+				}
+			}
```

Comment:
discussed with @afugazzotto offline that I would update this logic:  I was confused about the use of `!areDefFieldsMap` to mean "are they either not maps, or is one of them nil?"

Reply:
I do recognize that what I have is more verbose than the original, but I think it's okay because to me, it seems easier to understand.